File: src/modules/Pieces/Utils/HasMoved/HasMoved.cpp
#pragma once

#include "HasMoved.h"

HasMoved::HasMoved(PieceParams* params):
    Piece<RawPosition>(params),
    hasMoved(false)
{}

bool HasMoved::getHasMoved() {
    return this->hasMoved;
}

void HasMoved::onMoveEnd(int column, int row) {
    Piece::onMoveEnd(column, row);

    this->hasMoved = true;
}

File: packages/interactive_2d/chess/chess_react/src/Sketches/ChessSketch/elements/Pieces/Utils/HasMoved.ts
import { Piece } from "../Piece";
import { Position } from "../types";

export abstract class HasMoved extends Piece {
    private _hasMoved = false;

    public get hasMoved(): boolean {
        return this._hasMoved;
    }

    /**
     * @override
     */
    public onMoveEnd(rawPosition: Position): void {
        super.onMoveEnd(rawPosition);
        this._hasMoved = true;
    }
}
